### ISSUE TITLE
Database performance improvements

### ISSUE THREAD

**Issue by alvyxaz** [alvyxaz/barebones-masterserver#83]
Hey! 

I'm writing this to start a discussion

At the moment, one of the biggest bottlenecks within the framework is related to database accesses. It's not that bad when you have database relatively close (or on the same machine) to the master server, but if there's a 250ms ping between both of them - it can become a bit of a problem.

Of the top of my head, I think there are two viable options to solve this:

1. **Make database access asynchronous/with callbacks** (to support concurrency) - this is probably the easiest change, and should be made anyway
1. Load all of the database into RAM, so data can be accessed immediately - this might not be necessary, but could come as a separate "database interface implementation"

**If you have any thoughts on this - let me know! **

I'll probably work on this for the next update

**Comment by arcanor** [comment]
I agree on your point 1, but have some concerns that it may not be as easy to do as we might hope.

I disagree on point 2 (loading everything into memory), mostly because the database may become extremely large, and that's what SQL and the data server are already doing: managing the data.

Maybe this is off-topic, but maybe the master server shouldn't be the only way to get data. I have a feeling that with thousands of sessions going on simultaneously, as well messages sent through master via the ChatModule, the master server may become a bottleneck, and single point of failure.

Is it possible to offload specific modules to a separate server (i.e. a new kind of secure server, similar to a game server, but with special privileges)? Like maybe a 2nd database module that could point to game static data (not admin data), or maybe having the ChatModule handled by a different process?

**Comment by angusmf** [comment]
Async DB access please! And +1 for being able to run MS modules in separate/duplicate processes. I think I read your comments in the wiki somewhere about difficulties in 2-way syncing of profile data, and that it was possibly planned for later. I assume this would be a similar sort of problem, where there would be extra state and communication to keep the instances synchronized. If you've already put thought into this, would be interested to know what you've come up with.

**Comment by arcanor** [comment]
Note that I don't have any personal experience with this, I'm just talking off the top of my head! But I have concerns regarding transaction/commit capability.

As far as async database access goes, most relational database engines allow for transactions, but other types of engines may not. For example, MongoDB is not ACID compliant except when operating on a single document [http://stackoverflow.com/questions/27038391/transaction-support-in-mongodb](url), and probably most NoSQL engines are similar (including LiteDB). Without multi-document transactions, asynchronous compound read+write operations on highly active datasets may lose synchronization.

Maybe async could be added for RDBMSs only, as an optional feature?

**Comment by RikuTheFuffs** [comment]
Hi Guys, in my opinion the optimal solution would be avoiding managing the database directly. As @arcanor said, if everything relies on a single master server, the master server becomes a **huge** single point of failure, which means that it won't be scalable (and if you're using msf, scalability is something you probably need). Wouldn't it be easier to just have a separate REST API server and let the master server communicate with it only when it's really needed? (I.E: when you have to deal with requests that only an authenticated server should be able to do, such as giving exp to a player's account)
